cosmic-settings/src/pages/display/mod.rs

```diff
@@ -115,6 +129,30 @@ pub struct Page {
     cache: ViewCache,
     //  context: Option<ContextDrawer>,
     display_arrangement_scrollable: cosmic::widget::Id,
+    /// The setting to revert to if the next dialog page is cancelled, and
+    /// the instant the setting was changed.
+    dialog: Option<(Randr, time::Instant)>,
+}
+
+/// Causes a dialog to be shown. If the dialog fails, the revert_action will
+/// be performed.
+/// The revert_action should be which Randr call will revert the setting change.
+/// # Example
+/// ```
+/// fn set_scale(&mut self, scale: usize) -> Command<app::Message> {
+///
+///     // Sets the action to undo to the current value of the scale
+///     let revert_action = Randr::Scale(self.config.scale);
+///     set_dialog!(self, revert_action);
+///
+///     self.config.scale = scale;
+///     self.exec_randr(Randr::Scale(scale))
+/// }
+/// ```
+macro_rules! set_dialog {
+    ($self: ident, $revert_action: expr) => {
+        $self.dialog = Some(($revert_action, time::Instant::now()))
```

Comment:
I think it'd be better to make this a method of our Page

Reply:
I changed it to a method with the signature:
```rust
fn set_dialog(&mut self, revert_request: Randr, current_request: Randr) => Command<app::Message> {
```